entity-framework/core/testing/choosing-a-testing-strategy.md

```diff
@@ -0,0 +1,112 @@
+
+For information on how to use SQLite for unit testing, see the [unit testing page](xref:core/testing/unit-testing#sqlite-in-memory).
+
+### InMemory as a database fake
+
+As an alternative to SQLite, EF Core also comes with an InMemory provider. Although this provider was originally designed to support internal testing of EF Core itself, some developers use it as a database fake when unit testing EF Core using EF Core. Doing so is **highly discouraged**: as a database fake, InMemory has the same issues as SQLite (see above), but in addition has the following additional limitations:
+
+* The InMemory provider generally supports less query types than the SQLite provider, since it isn't a relational database. More queries will fail or behave differently in comparison to your production database.
+* Transactions are not supported.
+* Raw SQL is completely unsupported. Compare this with SQLite, where it's possible to use raw SQL, as long as that SQL works in the same way on SQLite and your production database.
+* The InMemory provider has not been optimized for performance, and will generally work slower than SQLite in in-memory mode.
+
+In summary, InMemory has all the disadvantages of SQLite, along with a few more - and offers no advantages in return. If you are looking for a simple, in-memory database fake, use SQLite instead of the InMemory provider; but consider using the repository pattern instead as described below.
+
+For information on how to use InMemory for unit testing, see the [unit testing page](xref:core/testing/unit-testing#inmemory).
+
+### Mocking or stubbing DbContext and DbSet
+
+This approach typically uses a mock framework to create a test double of `DbContext` and `DbSet`, and tests against those doubles. Mocking `DbContext` can be a good approach for testing various *non-query* functionality, such as calls to <xref:Microsoft.EntityFrameworkCore.DbContext.Add%2A> or <xref:Microsoft.EntityFrameworkCore.DbContext.SaveChanges>, allowing you to verify that your code called them in write scenarios.
+
+However, properly mocking `DbSet` *query* functionality is not possible, since queries are expressed via LINQ operators, which are static extension method calls over `IQueryable`. As a result, when some people talk about "mocking `DbSet`", what they really mean is that they stub `DbSet` out with an in-memory collection, and then evaluate query operators against that collection in memory, just like a simple `IEnumerable`. This isn't the same as actually mocking or stubbing the database, which would involve stubbing out query **end results**, rather than query inputs (i.e. the database table, as represented by a `DbSet`).
```

Comment:
```suggestion
However, properly mocking `DbSet` *query* functionality is extremely difficult, since queries are expressed via LINQ operators, which are static extension method calls over `IQueryable`. As a result, when some people talk about "mocking `DbSet`", what they really mean is that they stub `DbSet` out with an in-memory collection, and then evaluate query operators against that collection in memory, just like a simple `IEnumerable`. This isn't the same as actually mocking or stubbing the database, which would involve stubbing out query **end results**, rather than query inputs (i.e. the database table, as represented by a `DbSet`).
```

Reply:
Or maybe, "practically impossible". You can do it--you basically write your own LINQ provider for your tests.

Reply:
Yeah, as I wrote above - I think we're jumping between two definitions of what it means to mock DbSet/query.

I think most people understand this to mean having a DbSet backed by an in-memory collection, and applying Enumerable operators to it (technically faking rather than mocking). That's easy to do, but not really more useful than the in-memory provider.

The 2nd meaning (use a mock/fake LINQ provider) is sort-of what we provide with the in-memory provider; sure, we provide a fake EF provider rather than non-EF LINQ provider, but that's quite similar. Obviously users writing their own LINQ provider makes no sense, and I'm not sure what benefits that would bring over our in-memory provider.

Reply:
> Users writing their own LINQ provider makes no sense

Exactly.

> Obviously.

To you... :-D